components/sup/src/manager/service/supervisor.rs

```diff
@@ -167,14 +170,56 @@ impl Supervisor {
     }
 
     #[cfg(windows)]
-    fn user_info(&self, pkg: &Pkg) -> Result<UserInfo> {
+    fn user_info(&self, pkg: &Pkg, launcher: &LauncherCli) -> Result<UserInfo> {
+        // We have changed the implementation of get_current_username in core
+        // to use a win32 call GetUserNameW instead of the USERNAME environment
+        // variable. This introduces a problem if we are using an older launcher
+        // because both might get the current user name and an older launcher would
+        // get a different value if the current user is the SYSTEM user. The API
+        // call returns 'system' but the environment variable would be the host
+        // name followed by a dollar sign. So when an older launcher attempts to
+        // spawn a service and the supervisor is telling it to use the 'system' user
+        // name, the launcher will try to spawn the service using CreateProcessAsUserW
+        // instead of CreateProcessW thinking it needs to spawn the process as a
+        // different user. CreateProcessAsUserW will fail because it will attempt to
+        // authenticate the system user to create its access token which cannot be done
+        // with the system user.
+        //
+        // So we will check the version of the launcher. This version check is added
+        // in the same set of commits that changed the current username inspection
+        // implementation. So if the version handler is not supported, which it would
+        // not be in an older launcher, we will convert the user name to the old style
+        // system user using the host name.
+        let user = {
+            if pkg.svc_user == "system" {
+                let legacy_user = env::var("COMPUTERNAME")?.to_lowercase() + "$";
+                match launcher.version() {
+                    // 14221 is the last unstable launcher version as of the writing
+                    // of this comment.
+                    Ok(v) if v > 14221 => pkg.svc_user.clone(),
+                    Ok(_) => legacy_user,
+                    Err(err) => {
+                        match err {
+                            Timeout => {
+                                error!("Error getting version from launcher: {:?}", err);
+                                legacy_user
+                            }
+                            _ => return Err(Error::Launcher(err)),
+                        }
+                    }
```

Comment:
You could write this as:

```rust
Err(Timeout) => {
    error!("Error getting version from launcher: {:?}", err);
    legacy_user
}
Err(err) => {
    return Err(Error::Launcher(err));
}
```

to avoid an extra match statement.

Reply:
I actually struggled a bit with this and ended up with the seemingly verbose implementation I used. Your suggestion fails because `err` in `error!("Error getting version from launcher: {:?}", err);` is not in scope. Is there a way to specify the `Timeout` type and also get its value?

Reply:
Yep there is; sorry I missed that. A [binding](https://doc.rust-lang.org/stable/rust-by-example/flow_control/match/binding.html) should work. Something like:

```rust
Err(e @ Timeout) => {
    error!("Error getting version from launcher: {:?}", err);
    legacy_user
}
Err(err) => {
    return Err(Error::Launcher(err));
}
```

Reply:
ahh sweet. thanks!